### PR TITLE
refactored motd_formatter tests

### PR DESCRIPTION
**Pull request by Bubballoo3**
Response to #1540 to adjust tests to use MotdFile class and the formatter factory instead of calling formatters directly. Tests for missing files were modified to reflect a nil response from the factory, and tests where the MotdFile object was nil were removed because formatters should only ever be created by non-nil MotdFile instances.

### CODE REVIEW

File: apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb [hunk 1]
```diff
@@ -1,88 +1,65 @@
...
-    path = "#{Rails.root}/test/fixtures/files/motd_valid"
-    motd_file = MotdFile.new(path)
-    formatted_motd = MotdFormatterMarkdownErb.new(motd_file)
-    expected_file = OodAppkit.markdown.render(motd_file.content)
+    with_modified_env({ 'MOTD_FORMAT': "markdown_erb",'MOTD_PATH': "#{Rails.root}/test/fixtures/files/motd_valid" }) do
+      motd_file = MotdFile.new
+      expected_file = OodAppkit.markdown.render(motd_file.content)
 
-    assert_equal expected_file, formatted_motd.content
+      assert_equal expected_file, motd_file.formatter.content
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: I don't know if we should still continue to test against `OodAppkit.markdown.render(motd_file.content)`.  

Isn't this just `assert_equal '', motd_file.formatter.content`?

I don't think we should use a different algorithm to test against. Instead I think we should just use strings or fixture files. 

Put another way - this tests `are these 2 algorithms the same?`, when we should actually be asking `is the output of this algorithm (erb formatted content from an invalid file) equal to this constant?`

Comment (johrstrom) [reply apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: Errr sorry I'm mixing up valid an invalid, but I think my point still stands. I'd prefer to have a fixture file to test against rather than whatever `OodAppkit.markdown.render` returns.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb [hunk 2]
```diff
@@ -1,88 +1,65 @@
...
 
   test "motd-formatter-md-erb returns a standard error when given a invalid motd erb file" do
-    path = "#{Rails.root}/test/fixtures/files/motd_erb_standard_error"
-    motd_file = MotdFile.new(path)
-    
-    assert_raises(Exception) {
-      MotdFormatterMarkdownErb.new(motd_file)
-    }
+    with_modified_env({ 'MOTD_FORMAT': "markdown_erb",'MOTD_PATH': "#{Rails.root}/test/fixtures/files/motd_erb_standard_error" }) do
+      assert_raises(Exception) {
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: `Exception` is not a `StandardError` as the test case's name would have you believe. Seems like this case is off to begin with.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb [hunk 3]
```diff
@@ -1,88 +1,65 @@
...
-    expected_content = "<h1>Some Markdown file</h1>\n" \
-    "\n" \
-    "var msg = 'this was a script';\n"
-
-    assert_not_nil(content)
-    assert_equal(expected_content, content)
+    with_modified_env({ 'MOTD_FORMAT': "markdown_erb",'MOTD_PATH': "#{Rails.root}/test/fixtures/files/motd_md_erb_w_html" }) do
+      expected_content = "<h1>Some Markdown file</h1>\n" \
+      "\n" \
+      "var msg = 'this was a script';\n"
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: While we're at this, can you turn this into a HEREDOC just to make it a little cleaner?

File: apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb [hunk 4]
```diff
@@ -1,88 +1,65 @@
...
-    expected_content = "<h1>Some Markdown file</h1>\n" \
-    "\n" \
-    "<script>var msg = 'this was a script';</script>\n"
-
-    assert_not_nil(content)
-    assert_equal(expected_content, content)
+    with_modified_env({ 'MOTD_FORMAT': "markdown_erb",'MOTD_PATH': "#{Rails.root}/test/fixtures/files/motd_md_erb_w_html" }) do  
+      expected_content = "<h1>Some Markdown file</h1>\n" \
+      "\n" \
+      "<script>var msg = 'this was a script';</script>\n"
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: While we're at this, can you turn this into a HEREDOC just to make it a little cleaner?

File: apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb [hunk 5]
```diff
@@ -1,55 +1,41 @@
...
 
   test "plaintext-formatter-erb throws a standard error when given an invalid motd erb file" do
-    path = "#{Rails.root}/test/fixtures/files/motd_erb_standard_error"
-    motd_file = MotdFile.new(path)
-    
-    assert_raises(Exception) {
-      MotdFormatterPlaintextErb.new(motd_file)
-    }
+    with_modified_env({ 'MOTD_FORMAT': "text_erb", 'MOTD_PATH': "#{Rails.root}/test/fixtures/files/motd_erb_standard_error" }) do
+      assert_raises(Exception) {
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb]: Another `Exception` instead of `StandardError`.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb [hunk 6]
```diff
@@ -1,55 +1,41 @@
...
-    formatted_motd = MotdFormatterPlaintextErb.new(motd_file)
-
-    assert_equal '', formatted_motd.content
-  end
-
-  test "plaintext-formatter-erb returns an empty string when given nil" do
-    motd_file = nil
-    formatted_motd = MotdFormatterPlaintextErb.new(motd_file)
-
-    assert_not_nil formatted_motd.content
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb:52]: You seemed to have removed these 2 test cases. Are they handled elsewhere or what's the reasoning behind removing them?

File: apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_test.rb [hunk 7]
```diff
@@ -6,33 +6,22 @@ class MotdFormatterPlaintextTest < ActiveSupport::TestCase
...
-    formatted_motd = MotdFormatterPlaintext.new(motd_file)
-
-    assert_equal '', formatted_motd.content
-  end
-
-  test 'test when motd formatter_plaintext_nil' do
-    motd_file = nil
-    formatted_motd = MotdFormatterPlaintext.new(motd_file)
-
-    assert_not_nil formatted_motd.content
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_test.rb:36]: You seemed to have removed this test case against `nil`.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_rss_test.rb [hunk 8]
```diff
@@ -3,42 +3,26 @@
...
-    formatted_motd = MotdFormatterRss.new motd_file
-
-    assert_nil formatted_motd.content
-  end
-
-  test "test when motd formatter_rss_nil" do
-    motd_file = nil
-    formatted_motd = MotdFormatterRss.new motd_file
-
-    assert_nil formatted_motd.content
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_rss_test.rb:42]: Another test case against nil you seemed to have removed.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb (hunk 2 above)
Comment (Bubballoo3) [reply apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: It looks like the actual error that is generated is a NoMethodError, and NoMethodError < StandardError < Exception. I am guessing that they did intend for it to test StandardError, since exceptions under StandardError can be rescued. Of course since you can put any ruby code you like in a ERB file, I don't quite know that this is a strong guarantee that some other exception will not be encountered, and all we are observing is the error in the ERB bubbling up (rather than MotdFile issuing a new error when the ERB could not be read). But I imagine with normal use the vast majority of ERB errors would be some variety of StandardError, so maybe this is ok.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_rss_test.rb (hunk 8 above)
Comment (Bubballoo3) [reply apps/dashboard/test/models/motd_formatter/motd_formatter_rss_test.rb:42]: Those cases became unnecessary after the use of the formatter factory. They test behavior when the `MotdFile` object is nil, but if we only get formatters from `motd_file.formatter`, then we should only need to consider the cases where `motd_file` is an instance of `MotdFile`. Otherwise all we would be testing is that `nil.formatter` returns a NoMethodError

File: apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb (hunk 6 above)
Comment (Bubballoo3) [reply apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb:52]: I just changed the title of the first one to reflect the new testing, and the second one is a nil case that no longer applied.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb [hunk 9]
```diff
@@ -1,88 +1,69 @@
...
 class MotdFormatterMarkdownErbTest < ActiveSupport::TestCase
-  include MotdFormatter
   test "motd-formatter-md-erb returns valid motd file when given a valid motd file" do
-    path = "#{Rails.root}/test/fixtures/files/motd_valid"
-    motd_file = MotdFile.new(path)
-    formatted_motd = MotdFormatterMarkdownErb.new(motd_file)
-    expected_file = OodAppkit.markdown.render(motd_file.content)
+    with_modified_env({ 'MOTD_FORMAT': "markdown_erb",'MOTD_PATH': "#{Rails.root}/test/fixtures/files/motd_valid" }) do
+      motd_file = MotdFile.new
+      expected_file = File.read "#{Rails.root}/test/fixtures/files/motd_valid_html"
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: This test initially used `motd_valid` but now uses `motd_valid_html`?

What's more, this test is supposed to test an `.md.erb` file not an `.html` file.

Comment (johrstrom) [reply apps/dashboard/test/models/motd_formatter/motd_formatter_markdown_erb_test.rb]: 🤦‍♂️ I see what this is now, disregard the initial comment.

Can you use parenthesis here for this `File.read` call?

File: apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb [hunk 10]
```diff
@@ -1,55 +1,41 @@
 require 'test_helper'
 
 class MotdFormatter::PlaintextErbTest < ActiveSupport::TestCase
-  include MotdFormatter
   test "plaintext-formatter-erb returns a valid motd file when given a valid motd file" do
     path = "#{Rails.root}/test/fixtures/files/motd_valid"
-    motd_file = MotdFile.new(path)
-    formatted_motd = MotdFormatterPlaintextErb.new(motd_file)
-    expected_file = File.open(path).read
+    with_modified_env({ 'MOTD_FORMAT': "text_erb", 'MOTD_PATH': path }) do
```
Comment (johrstrom) [inline apps/dashboard/test/models/motd_formatter/motd_formatter_plaintext_erb_test.rb:6]: The initial test was plain text format. Indeed this class name also reinforces that - `PlaintextErbTest`.

File: apps/dashboard/test/models/motd_formatter/motd_formatter_rss_test.rb (hunk 8 above)
Comment (johrstrom) [reply apps/dashboard/test/models/motd_formatter/motd_formatter_rss_test.rb:42]: OK.
